torch/overrides.py

```diff
@@ -1819,6 +1817,16 @@ def __init__(self):
     def __torch_function__(self, func, types, args=(), kwargs=None):
         raise NotImplementedError()
 
+    def __enter__(self):
+        if hasattr(self, "inner"):
```

Comment:
Flagging again that in the next PR in the stack, we remove the ability to set "inner" through the constructor which limits the errors that can happen here. In hindsight, I like that we set inner here instead of in the constructor because before this the following code is valid:

```
inner_mode = MyMode(inner=None)
with enable_torch_dispatch_mode(MyMode(inner=inner_mode)):
  ...
```

which is really weird because it will dispatch to `inner_mode` even though it's never used in a context manager. By setting it in enter, users will no longer be able to have this in their code

Reply:
This is actually intentional; in functorch there are circumstances (custom vjp) where you close over some mode state and want to reapply it at some later point: you want to apply all the modes here. So the raw setter API let you apply as many modes as you want at once. I have never personally used it for anything though!

Reply:
I think the code as written here is correct however! __enter__ should not be OK for restoring a mode; we can have a dedicated method for that